### DISCUSSION TITLE
Pausing & Resuming a charging session

### DISCUSSION THREAD

**Discussion by GrantKeymer** [lbbrhzn/ocpp#333] [Q&A] [answered]
Now I've got most of my Dashboard operating, I realised there is still one function I'm missing compared to the Wallbox App...

**The ability to Pause and Resume a charging session without starting a new one.**

It would be nice to do this because the session values for kWh and Time are reset to zero when a new session starts.  Suppose I have a target kWh that I want to reach, rather than just letting the charger run until the EV is full.  If the session can be paused when insufficient solar power is available and then resumed once the clouds have cleared, it would be preferable to having many short sessions to accomplish the same thing.

Looking through the OCPP 1.6 spec, I don't see any specific provision for the Central System to pause or resume a charging session.  The Charge Point itself can initiate such an action by setting its status to EVSE Suspended.

Is there some way to work around this so I can effectively force the Charge Point to change to status EVSE Suspended?

**Comment by drc38** [comment]
Just set your max charging current to 0, then back to say 16A to resume

**Comment by GrantKeymer** [comment, marked as answer]
I've just tried that using the slider in HA.  Sure enough, the log file shows the Max. Current setting of zero being sent out, followed by "Accepted" from the charger.  However, the next lot of Meter Readings from the charger show it's set to 6 Amps and it stays that way.

Good idea, but no cigar 😕

**Reply by drc38** [reply]
Not sure then, it works on mine. Have you checked whether you can alter the 6A setting via a configuration key?

**Reply by GrantKeymer** [reply]
There isn't any such key available in the Wallbox app.  Where else might I find it?

**Comment by drc38** [comment]
Charger manual on ocpp or using the get configuration service. If you don't send a key I think it will return all keys.

**Reply by GrantKeymer** [reply]
I've checked right through the Wallbox Document "OCPP Activation Guide" and there's nothing about configuration keys.

I presume when you say use the "Get Configuration Service" that would entail modifying my local copy of the existing code somehow?

**Reply by drc38** [reply]
No under Developer Tools | Services in HA

**Reply by drc38** [reply]
The main keys are explained in the Ocpp specification, but the charger can also have custom keys set by the manufacturer

**Reply by GrantKeymer** [reply]
Aha, I finally found it...

**Open Charge Point Protocol (OCPP): Get configuration values for charger
service: ocpp.get_configuration**

**All available parameters**

Below that is blank apart from some example data.  It doesn't seem possible to get a list of all existing keys from the charger  
by leaving the key field blank.  Unless I put a key in there, the 'Call Service' button stays disabled.

I tried a couple of keys from the OCPP spec.  The first one gave me this...

**2022-01-27 23:39:07 INFO (MainThread) [ocpp] : receive message [3,"8d53b069-2399-4a05-a87b-4d39833d79a0",
{
"unknownKey": [],
"configurationKey": [
{
"readonly": false,
"key": "MeterValueSampleInterval",
"value": "60"
}
]
}]**

And another gave me this...

**2022-01-27 23:41:51 INFO (MainThread) [ocpp] : receive message [3,"e5a4e60d-da1a-4540-b58c-4cebc9bfa4f2",
{
"unknownKey": [],
"configurationKey": [
{
"readonly": true,
"key": "SupportedFeatureProfiles",
"value": "Core,FirmwareManagement,LocalAuthListManagement,SmartCharging,RemoteTrigger"
}
]
}]**

I don't see anything concerning Minimum Charging Current 😕

**Reply by drc38** [reply]
Use `ocpp_key: ""` to get all keys

**Reply by lbbrhzn** [reply]
```
022-01-27 22:34:53 INFO (MainThread) [ocpp] pulsar: send [2,"1ad349cb-3b70-4ed5-b726-bda24db81a7d","GetConfiguration",{}]
2022-01-27 22:34:53 INFO (MainThread) [ocpp] pulsar: receive message [3,"1ad349cb-3b70-4ed5-b726-bda24db81a7d",
{
"unknownKey": [],
"configurationKey": [
{
"readonly": false,
"key": "AllowOfflineTxForUnknownId",
"value": "false"
},

{
"readonly": false,
"key": "AuthorizationCacheEnabled",
"value": "true"
},

{
"readonly": false,
"key": "AuthorizeRemoteTxRequests",
"value": "false"
},

{
"readonly": false,
"key": "BlinkRepeat",
"value": "10"
},

{
"readonly": false,
"key": "ClockAlignedDataInterval",
"value": "900"
},

{
"readonly": false,
"key": "ConnectionTimeOut",
"value": "120"
},

{
"readonly": false,
"key": "ConnectorPhaseRotation",
"value": "1.Unknown"
},

{
"readonly": true,
"key": "ConnectorPhaseRotationMaxLength",
"value": "1"
},

{
"readonly": true,
"key": "ConfigurationMaxKeys",
"value": "50"
},

{
"readonly": false,
"key": "HeartbeatInterval",
"value": "3600"
},

{
"readonly": false,
"key": "LightIntensity",
"value": "100"
},

{
"readonly": false,
"key": "LocalAuthorizeOffline",
"value": "true"
},

{
"readonly": false,
"key": "LocalPreAuthorize",
"value": "false"
},

{
"readonly": false,
"key": "MaxEnergyOnInvalidId",
"value": "0"
},

{
"readonly": false,
"key": "MeterValuesAlignedData",
"value": "Energy.Active.Import.Register"
},

{
"readonly": true,
"key": "MeterValuesAlignedDataMaxLength",
"value": "10"
},

{
"readonly": false,
"key": "MeterValuesSampledData",
"value": "Energy.Active.Import.Register,Energy.Reactive.Import.Register,Energy.Active.Import.Interval,Energy.Reactive.Import.Interval,Power.Active.Import,Power.Reactive.Import,Power.Offered,Power.Factor,Current.Import,Current.Offered,Voltage,Frequency,RPM,SoC,Temperature,Current.Export,Energy.Active.Export.Register,Energy.Reactive.Export.Register,Energy.Active.Export.Interval,Energy.Reactive.Export.Interval,Power.Active.Export,Power.Reactive.Export"
},

{
"readonly": true,
"key": "MeterValuesSampledDataMaxLength",
"value": "10"
},

{
"readonly": false,
"key": "MeterValueSampleInterval",
"value": "60"
},

{
"readonly": false,
"key": "MinimumStatusDuration",
"value": "1"
},

{
"readonly": true,
"key": "NumberOfConnectors",
"value": "1"
},

{
"readonly": false,
"key": "ResetRetries",
"value": "0"
},

{
"readonly": false,
"key": "StopTransactionOnEVSideDisconnect",
"value": "true"
},

{
"readonly": false,
"key": "StopTransactionOnInvalidId",
"value": "true"
},

{
"readonly": false,
"key": "StopTxnAlignedData",
"value": ""
},

{
"readonly": true,
"key": "StopTxnAlignedDataMaxLength",
"value": "0"
},

{
"readonly": false,
"key": "StopTxnSampledData",
"value": ""
},

{
"readonly": true,
"key": "StopTxnSampledDataMaxLength",
"value": "0"
},

{
"readonly": true,
"key": "SupportedFeatureProfiles",
"value": "Core,FirmwareManagement,LocalAuthListManagement,SmartCharging,RemoteTrigger"
},

{
"readonly": true,
"key": "SupportedFeatureProfilesMaxLength",
"value": "5"
},

{
"readonly": false,
"key": "TransactionMessageAttempts",
"value": "3"
},

{
"readonly": false,
"key": "TransactionMessageRetryInterval",
"value": "60"
},

{
"readonly": false,
"key": "UnlockConnectorOnEVSideDisconnect",
"value": "true"
},

{
"readonly": false,
"key": "WebSocketPingInterval",
"value": "60"
},

{
"readonly": false,
"key": "LocalAuthListEnabled",
"value": "false"
},

{
"readonly": true,
"key": "LocalAuthListMaxLength",
"value": "5000000"
},

{
"readonly": true,
"key": "SendLocalListMaxLength",
"value": "1000000"
},

{
"readonly": true,
"key": "ReserveConnectorZeroSupported",
"value": "false"
},

{
"readonly": true,
"key": "ChargeProfileMaxStackLevel",
"value": "0"
},

{
"readonly": true,
"key": "ChargingScheduleAllowedChargingRateUnit",
"value": "Current"
},

{
"readonly": true,
"key": "ChargingScheduleMaxPeriods",
"value": "1"
},

{
"readonly": true,
"key": "ConnectorSwitch3to1PhaseSupported",
"value": "false"
},

{
"readonly": true,
"key": "MaxChargingProfilesInstalled",
"value": "1"
},

{
"readonly": false,
"key": "chargingALimitConn1",
"value": "32"
},

{
"readonly": false,
"key": "AuthEnabled",
"value": "true"
},

{
"readonly": false,
"key": "AuthDisabledIdTag",
"value": "NoAuthorization"
} 
]
}]

```

**Reply by lbbrhzn** [reply]
This might be an interesting key, @GrantKeymer 
```
{
"readonly": false,
"key": "chargingALimitConn1",
"value": "32"
}
```

**Reply by GrantKeymer** [reply]
Thanks guys, now I understand.  Will have a play with the key "chargingALimitConn1" later when I get home.

**Reply by GrantKeymer** [reply]
I can now confirm that the key "chargingALimitConn1" is present in the list of keys from my Pulsar Plus.

We have a service "GetConfiguration.req" on page 65 of the OCPP Spec.

It appears that I need to call the "ChangeConfiguration.req" function in order to change this key.  But I don't see this function listed in the available services provided by the OCPP integration.  Has it been implemented yet?

Or is there some other way to tinker with this function and see if it will allow me to Pause/Resume before writing any code?

**Reply by drc38** [reply]
The sevice is simply called `configure` in the integration

**Reply by GrantKeymer** [reply]
OK, now we have the answer...

```
2022-01-28 18:17:02 INFO (MainThread) [ocpp] : send [2,"fcc6ed9a-85e4-448c-9aaa-f819d3ce7d6b","ChangeConfiguration",{"key":"chargingALimitConn1","value":"0"}]
2022-01-28 18:17:02 DEBUG (MainThread) [websockets.server] > TEXT '[2,"fcc6ed9a-85e4-448c-9aaa-f819d3ce7d6b","Chan...mitConn1","value":"0"}]' [106 bytes]
2022-01-28 18:17:02 DEBUG (MainThread) [websockets.server] < TEXT '[3,"fcc6ed9a-85e4-448c-9aaa-f819d3ce7d6b",\n{\n"status": "Rejected"\n}]' [68 bytes]
2022-01-28 18:17:02 INFO (MainThread) [ocpp] : receive message [3,"fcc6ed9a-85e4-448c-9aaa-f819d3ce7d6b",
{
"status": "Rejected"
}]
2022-01-28 18:17:02 WARNING (MainThread) [custom_components.ocpp] Rejected while setting chargingALimitConn1 to 0

```
Setting the current limit to 0 is rejected, as is setting it to 5...

```
2022-01-28 18:17:32 INFO (MainThread) [ocpp] : send [2,"b356bbf8-88d4-483a-9e65-a4b47b96ab32","ChangeConfiguration",{"key":"chargingALimitConn1","value":"5"}]
2022-01-28 18:17:32 DEBUG (MainThread) [websockets.server] > TEXT '[2,"b356bbf8-88d4-483a-9e65-a4b47b96ab32","Chan...mitConn1","value":"5"}]' [106 bytes]
2022-01-28 18:17:32 DEBUG (MainThread) [websockets.server] < TEXT '[3,"b356bbf8-88d4-483a-9e65-a4b47b96ab32",\n{\n"status": "Rejected"\n}]' [68 bytes]
2022-01-28 18:17:32 INFO (MainThread) [ocpp] : receive message [3,"b356bbf8-88d4-483a-9e65-a4b47b96ab32",
{
"status": "Rejected"
}]
2022-01-28 18:17:32 WARNING (MainThread) [custom_components.ocpp] Rejected while setting chargingALimitConn1 to 5

```
In each of those cases, the charger returns a status indicating that "chargingALimitConn1" remains at 6...

```
2022-01-28 18:17:47 DEBUG (MainThread) [websockets.server] < TEXT '[3,"ee3870df-3639-4cf5-9b23-68d3c69b6a93",\n{\n..."value": "6"\n} \n]\n}]' [156 bytes]
2022-01-28 18:17:47 INFO (MainThread) [ocpp] : receive message [3,"ee3870df-3639-4cf5-9b23-68d3c69b6a93",
{
"unknownKey": [],
"configurationKey": [
{
"readonly": false,
"key": "chargingALimitConn1",
"value": "6"
}
]
}]
```

If I send the same ChangeConfiguration.req with a value of 6, 7 or 10 it is accepted.  So it seems, I am no closer to achieving my goal 😞

**Reply by drc38** [reply]
If you have no luck with the charger ovms can be used to pause/resume from the car's end.

**Comment by GrantKeymer** [comment]
Here's an update on what happens if I deliberately try to set Max_Current_Offered above the range of the charger.  In my case, Wallbox Pulsar Plus has an upper limit of 32 Amps.  The latest version of OCPP Integration now allows Maximum Current to be configured so I chose 40 Amps.  Then restarted HA and plugged in my EV.  Initially I started charging at 11 Amps just to let everything settle down in the normal working range.  Here's what happened when I selected a 40 Amp charging rate...

```
2022-02-18 09:53:28 INFO (MainThread) [ocpp] : send [2,"3d5f81b7-f223-4d31-9c38-b7d1e85259af","SetChargingProfile",{"connectorId":0,"csChargingProfiles":{"chargingProfileId":8,"stackLevel":0,"chargingProfileKind":"Relative","chargingProfilePurpose":"ChargePointMaxProfile","chargingSchedule":{"chargingRateUnit":"A","chargingSchedulePeriod":[{"startPeriod":0,"limit":40.0}]}}}]
2022-02-18 09:53:28 DEBUG (MainThread) [websockets.server] > TEXT '[2,"3d5f81b7-f223-4d31-9c38-b7d1e85259af","SetC...d":0,"limit":40.0}]}}}]' [325 bytes]
2022-02-18 09:53:28 DEBUG (MainThread) [websockets.server] < TEXT '[3,"3d5f81b7-f223-4d31-9c38-b7d1e85259af",\n{\n"status": "Accepted"\n}]' [68 bytes]
2022-02-18 09:53:28 INFO (MainThread) [ocpp] : receive message [3,"3d5f81b7-f223-4d31-9c38-b7d1e85259af",
{
"status": "Accepted"
}]
```

So that looked fine.  However, the next lot of Meter Values showed this...

`2022-02-18 09:53:34 INFO (MainThread) [ocpp] : receive message [2,"aca3ba02-465b-4086-9063-23af5443e160","MeterValues",{"meterValue": [{"timestamp": "2022-02-17T20:53:32Z","sampledValue": [{"value": "70","format": "Raw","location": "Outlet","context": "Sample.Periodic","measurand": "Energy.Active.Import.Register","unit": "Wh"},{"value": "3019","format": "Raw","location": "Outlet","context": "Sample.Periodic","measurand": "Power.Active.Import","unit": "W"},{"value": "22170","format": "Raw","location": "Outlet","context": "Sample.Periodic","measurand": "Power.Offered","unit": "W"},{"value": "1e+01","format": "Raw","location": "Outlet","context": "Sample.Periodic","phase": "L1","measurand": "Current.Import","unit": "A"},{"value": "32","format": "Raw","location": "Outlet","context": "Sample.Periodic","measurand": "Current.Offered","unit": "A"},{"value": "222.000000","format": "Raw","location": "Outlet","context": "Sample.Periodic","phase": "L1-N","measurand": "Voltage","unit": "V"},{"value": "50.000000","format": "Raw","location": "Outlet","context": "Sample.Periodic","measurand": "Frequency"} ]} ],"connectorId": 1,"transactionId": 1645131096}]`

So the charger has restricted the Max Current Offered to 32 Amps, in accordance with its internal limits.  Unfortunately this still doesn't allow me to Pause/Resume the charging session but it was worth a try.
